src/components/Privacy/createComponent.js

```diff
@@ -0,0 +1,68 @@
+/*
+Copyright 2020 Adobe. All rights reserved.
+This file is licensed to you under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License. You may obtain a copy
+of the License at http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software distributed under
+the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
+OF ANY KIND, either express or implied. See the License for the specific language
+governing permissions and limitations under the License.
+*/
+
+import { assign } from "../../utils";
+
+export default ({
+  readStoredConsent,
+  taskQueue,
+  defaultConsent,
+  consent,
+  sendSetConsentRequest,
+  validateSetConsentOptions
+}) => {
+  const consentByPurpose = assign({}, defaultConsent, readStoredConsent());
+  consent.setConsent(consentByPurpose);
+
+  const readCookieIfQueueEmpty = () => {
+    if (taskQueue.length === 0) {
+      // Only read cookies when there are no outstanding setConsent
+      // requests. This helps with race conditions.
+      consent.setConsent(readStoredConsent());
+    }
+  };
+
+  return {
+    commands: {
+      setConsent(options) {
+        const validatedOptions = validateSetConsentOptions(options);
+        consent.suspend();
+        return taskQueue
+          .addTask(() => sendSetConsentRequest(validatedOptions))
+          .catch(error => {
+            readCookieIfQueueEmpty();
+            if (
+              error &&
+              error.message &&
+              error.message.includes("User is opted out")
+            ) {
+              throw new Error(
+                "The user previously declined consent, which cannot be changed."
+              );
+            }
+            throw error;
+          })
+          .then(readCookieIfQueueEmpty);
+      }
+    },
+    lifecycle: {
+      // Read the cookie here too because the consent cookie may change on any request
+      onResponse: readCookieIfQueueEmpty,
+      // Even when we get a failure HTTP status code, the consent cookie can
+      // still get updated. This could happen, for example, if the user is
+      // opted out in AudienceManager, but no consent cookie exists on the
+      // client. The request will be sent and the server will respond with a
+      // 403 Forbidden and a consent cookie.
+      onResponseFailure: readCookieIfQueueEmpty
```

Comment:
@Aaronius If we rename this one to `onResponseFailure` which might make sense, what would a use case for `onRequestFailure` be?

Reply:
There might not be one. One that might exist, which we're not doing right now, is that if the initial `interact` request goes out and other requests are queued up, but the request fails (e.g., no network connection), then we might want to reject all the commands that were waiting for that initial `interact` request to succeed. Currently, if the initial `interact` request fails or the response fails, I believe the queued commands just stay queued, for better or worse.

If we only went with an `onResponseFailure` lifecycle method, the question is whether it should be called when the request doesn't actually make it to the server. Probably not.

Reply:
Good catch, I will change this to onRequestFailure.  I like onRequestFailure for when we don't get a response, and onResponse should just handle any response from the server regardless of the response code.  My reasoning is that even when Konductor returns a non-200 status code, it will still return responses.  I'm trying to make component code as simple as possible, and I can't think of a case where the component would care.  I'm just thinking out loud here though. We would need to look at all the possible Konductor responses and see what make the most sense.

Reply:
> My reasoning is that even when Konductor returns a non-200 status code, it will still return responses.

As far as I'm aware, this is true as long as the request actually makes it to Konductor and Konductor is functioning properly. If JAG encounters an error, then it will respond with a non-200 status code and a non-JSON body describing the error. We could say that if we ever get a non-JSON response, we consider it a "request failure".

Reply:
Sorry, let me clarify. `onResponse` is always handed a `response` object, which I think makes sense. I think there are a few scenarios then that we need to manage and here's how I think we can deal with them:

(1) A response body that contains a regular Konductor response payload (JSON). For this, we can create a response object and call `onResponse`.
(2) An empty response body with an HTTP status code of 204 No Content. For this, we can still create a response object (it just won't have content in it) and still call `onResponse`.
(3) Any other response. This case is unexpected, and therefore I don't feel like we should attempt to create a response object or treat it in any way as successful. I think in this case we would call `onRequestFailure`.